### PR TITLE
Fix reconnect on refused connection

### LINKED ISSUE #1356: Reconnecting WebSocketService after a close call
I'm wondering how to reconnect a WebSocketService after a close without re-instantiating the WebSocketService object and the related Web3j object and contract.

### PR DESCRIPTION
**Pull request by zilm13**
### What does this PR do?
Toggles `shouldReConnect` for connection refused case too.
Without it I get something like this in logs when trying to reconnect node 2nd time after 1st attempt got connection refused:
```
Caused by: java.lang.IllegalStateException: WebSocketClient objects are not reuseable
at org.java_websocket.client.WebSocketClient.connect(WebSocketClient.java:372) ~[Java-WebSocket-1.5.2.jar:?]
at org.java_websocket.client.WebSocketClient.connectBlocking(WebSocketClient.java:386) ~[Java-WebSocket-1.5.2.jar:?]
at org.web3j.protocol.websocket.WebSocketService.connectToWebSocket(WebSocketService.java:129) ~[core-4.8.9.jar:?]
at org.web3j.protocol.websocket.WebSocketService.connect(WebSocketService.java:117) ~[core-4.8.9.jar:?]
``` 

### Where should the reviewer start?
`WebSocketClient.java`

### Why is it needed?
Otherwise user needs to recreate `WebSocketClient` and `WebSocketService` instances for every reconnection attempt when connection is refused.
Also will fix #1356 I guess

### CODE REVIEW

File: core/src/main/java/org/web3j/protocol/websocket/WebSocketService.java [hunk 1]
```diff
@@ -128,11 +128,10 @@ private void connectToWebSocket() throws InterruptedException, ConnectException
                         ? webSocketClient.reconnectBlocking()
                         : webSocketClient.connectBlocking();
 
+        shouldReConnect = true;
```
Comment (andrii-kl) [inline core/src/main/java/org/web3j/protocol/websocket/WebSocketService.java:131]: I would add a functionality to make some limited attempts of web socket connection. (webSocketClient.connectBlocking())

Comment (zilm13) [reply core/src/main/java/org/web3j/protocol/websocket/WebSocketService.java:131]: > I would add a functionality to make some limited attempts of web socket connection. (webSocketClient.connectBlocking())

I don't understand. `connect`which calls `connectToWebsocket` is a public API, so the external user could decide whether and when he wants to reconnect. You want to make such logic to be internal?

Comment (zilm13) [reply core/src/main/java/org/web3j/protocol/websocket/WebSocketService.java:131]: @andrii-kl added test, which fails in a current master

File: core/src/test/java/org/web3j/protocol/websocket/WebSocketServiceTest.java [hunk 2]
```diff
@@ -121,6 +122,21 @@ public void testThrowExceptionIfConnectionFailed() throws Exception {
...
+        assertThrows(
+                ConnectException.class,
+                () -> {
+                    service.connect();
+                });
+        service.connect();
+        // reconnectBlocking() should be called if 1st attempt was failed, if we call
+        // connectBlocking() for a second time, we will get IllegalStateException with real
+        // webSocketClient
+        verify(webSocketClient, times(1)).reconnectBlocking();
```
Comment (gtebrean) [inline core/src/test/java/org/web3j/protocol/websocket/WebSocketServiceTest.java:137]:
```suggestion
        verify(webSocketClient, atMostOnce()).reconnectBlocking();
```
nit, but to be consistent with other test cases

File: core/src/main/java/org/web3j/protocol/websocket/WebSocketService.java (hunk 1 above)
Comment (gtebrean) [reply core/src/main/java/org/web3j/protocol/websocket/WebSocketService.java:131]: @zilm13 I guess @andrii-kl wanted to say to set up a counter in order to establish a connection pool. I took a look at code and I don't think is necessary, even more there is the onClose() method that is called on every listener.
Your test makes sense. 👍

**Review by gtebrean — APPROVED** [review]

File: core/src/test/java/org/web3j/protocol/websocket/WebSocketServiceTest.java (hunk 2 above)
Comment (gtebrean) [reply core/src/test/java/org/web3j/protocol/websocket/WebSocketServiceTest.java]: I prefer times too, but also I like classes to be as clear and consistent as possible, so adding that now would mean to update all checks in test and is not really the case. Thanks for update!

**Review by andrii-kl — APPROVED** [review]

**Comment by zilm13** [comment]
@andrii-kl any schedule, when this could hit release? We need this change to implement Websockets support in Teku

**Pull request merged**